### PR TITLE
[Merged by Bors] - feat: add results for the computation of the volume of Zspan.fundamentalDomain

### PR DESCRIPTION
**Pull request by xroblot**
We prove that the covolume of a $\mathbb{Z}$-lattice of $\mathbb{R}^n$ generated by the basis $b$ is equal to the absolute value of the determinant of the matrix formed by the vectors of $b$, that is
```lean
theorem volume_fundamentalDomain [Fintype ι] [DecidableEq ι] (b : Basis ι ℝ (ι → ℝ)) :
     volume (fundamentalDomain b) = ENNReal.ofReal |(Matrix.of b).det|
```

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/Zlattice.lean [hunk 1]
```diff
@@ -300,6 +315,28 @@ protected theorem isAddFundamentalDomain [Finite ι] [MeasurableSpace E] [OpensM
...
+    simp [Basis.toMatrix_apply, LinearMap.toMatrix_apply, LinearEquiv.coe_coe, Basis.equiv_apply]
+
+theorem volume_fundamentalDomain [Fintype ι] [DecidableEq ι] (b : Basis ι ℝ (ι → ℝ)) :
+    volume (fundamentalDomain b) = ENNReal.ofReal |(Matrix.of b).det| := by
+  rw [measure_fundamentalDomain b volume (b₀ := Pi.basisFun ℝ ι),
+    show fundamentalDomain (Pi.basisFun ℝ ι) = Set.pi Set.univ fun _ : ι => Set.Ico (0 : ℝ) 1 by
+      ext; simp only [mem_fundamentalDomain, Pi.basisFun_repr, Set.mem_Ico, Set.mem_pi,
+        Set.mem_univ, forall_true_left],
+    volume_pi, Measure.pi_pi, Real.volume_Ico, sub_zero, ENNReal.ofReal_one, Finset.prod_const_one,
+    mul_one, ← Matrix.det_transpose]
```
Comment (ocfnash) [inline Mathlib/Algebra/Module/Zlattice.lean:337]:
```suggestion
  rw [measure_fundamentalDomain b volume (b₀ := Pi.basisFun ℝ ι), fundamentalDomain_pi_basisFun,
    volume_pi, Measure.pi_pi, Real.volume_Ico, sub_zero, ENNReal.ofReal_one, Finset.prod_const_one,
    mul_one, ← Matrix.det_transpose, Basis.coePiBasisFun.toMatrix_eq_transpose,
    Matrix.transpose_transpose]
```

File: Mathlib/Algebra/Module/Zlattice.lean [hunk 2]
```diff
@@ -300,6 +315,28 @@ protected theorem isAddFundamentalDomain [Finite ι] [MeasurableSpace E] [OpensM
...
+  have : FiniteDimensional ℝ E := FiniteDimensional.of_fintype_basis b
+  convert μ.addHaar_preimage_linearEquiv (b.equiv b₀ (Equiv.refl ι)) (fundamentalDomain b₀)
+  · rw [Set.eq_preimage_iff_image_eq (LinearEquiv.bijective _), map_fundamentalDomain,
+      Basis.map_equiv, Equiv.refl_symm, Basis.reindex_refl]
+  · rw [← LinearMap.det_toMatrix b₀, Basis.equiv_symm, Equiv.refl_symm]
+    congr
+    ext
+    simp [Basis.toMatrix_apply, LinearMap.toMatrix_apply, LinearEquiv.coe_coe, Basis.equiv_apply]
+
+theorem volume_fundamentalDomain [Fintype ι] [DecidableEq ι] (b : Basis ι ℝ (ι → ℝ)) :
```
Comment (ocfnash) [inline Mathlib/Algebra/Module/Zlattice.lean:330]:
```suggestion
@[simp]
theorem volume_fundamentalDomain [Fintype ι] [DecidableEq ι] (b : Basis ι ℝ (ι → ℝ)) :
```

**Comment by xroblot** [comment]
@ocfnash Thanks for the review. One of my frustration with this PR is that the same result is true for [Basis.parallelepiped](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Haar/OfBasis.html#Basis.parallelepiped), that is: 
```lean
example [Fintype ι] [DecidableEq ι] (b : Basis ι ℝ (ι → ℝ)) :
    volume (b.parallelepiped) = ENNReal.ofReal |(Matrix.of b).det|
```
but I was not able to find a nice way to deduce one from the other. Indeed, the quickest way I found to prove the result for `Basis.parallelepiped` is to redo essentially the proof (so I saved this result for another PR if needed). 

More generally, I think it is still lacking a good mechanism to connect `Basis.parallelepiped` and `Zspan.fundamentalDomain`...

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**